xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWiki.java

```diff
@@ -2756,20 +2757,28 @@ public String getUserPreference(String prefname, boolean useCookie, XWikiContext
     }
 
     /**
-     * First try to find the current locale in use from the XWiki context. If none is used and if the wiki is not
-     * multilingual use the default locale defined in the XWiki preferences. If the wiki is multilingual try to get the
-     * locale passed in the request. If none was passed try to get it from a cookie. If no locale cookie exists then use
-     * the user default locale and barring that use the browser's "Accept-Language" header sent in HTTP request. If none
-     * is defined use the default locale.
+     * Set the locale in the given context.
+     * <p>
+     * If {@code forceSupported} is true, then the locale will be set only if it is in the {@see availableLocales}. Note
+     * that all the parent locales are checked.
      *
-     * @return the locale to use
-     * @since 8.0M1
+     * @param locale the locale to use
+     * @param context the context
+     * @param availableLocales the accepted locales. Used only if {@see forceSupported} is true
+     * @param forceSupported determine if the {@see locale} should be checked against the {@see availableLocales}
+     * @return the locale that has been set or null
      */
-    public Locale getLocalePreference(XWikiContext context)
+    private Locale setLocale(Locale locale, XWikiContext context, Set<Locale> availableLocales, boolean forceSupported)
```

Comment:
If the passe Locale is completely different from the allowed locales you will end up with a null Locale. Shouldn't it fallback on default Locale instead ? Also you should should stop at Locale.ROOT, not null.